src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.StringSegment.cs

```diff
@@ -0,0 +1,496 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text.Json
+{
+    public sealed partial class Utf8JsonWriter
+    {
+        /// <summary>
+        /// Writes the text value segment as a partial JSON string.
+        /// </summary>
+        /// <param name="value">The value to write.</param>
+        /// <param name="isFinalSegment">Indicates that this is the final segment of the string.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the specified value is too large.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if this would result in invalid JSON being written (while validation is enabled).
+        /// </exception>
+        /// <remarks>
+        /// The value is escaped before writing.
+        /// </remarks>
+        public void WriteStringValueSegment(ReadOnlySpan<char> value, bool isFinalSegment)
+        {
+            JsonWriterHelper.ValidateValue(value);
+
+            if (!_options.SkipValidation && _tokenType != Utf8JsonWriter.StringSegmentSentinel)
+            {
+                ValidateWritingValue();
+            }
+
+            if (_tokenType != Utf8JsonWriter.StringSegmentSentinel)
+            {
+                WriteStringSegmentProlog();
+                _tokenType = Utf8JsonWriter.StringSegmentSentinel;
+            }
+
+            // The steps to write a string segment are to complete the previous partial code point
+            // and escape either of which might not be required so there is a fast path for each of these steps.
+            if (HasPartialCodePoint)
+            {
+                WriteStringSegmentWithLeftover(value, isFinalSegment);
+            }
+            else
+            {
+                WriteStringSegmentEscape(value, isFinalSegment);
+            }
+
+            if (isFinalSegment)
+            {
+                WriteStringSegmentEpilog();
+                SetFlagToAddListSeparatorBeforeNextItem();
+                _tokenType = JsonTokenType.String;
+            }
+        }
+
+        private void WriteStringSegmentWithLeftover(scoped ReadOnlySpan<char> value, bool isFinalSegment)
+        {
+            Span<char> combinedBuffer = stackalloc char[2];
+
+            GetPartialUtf16CodePoint(out scoped ReadOnlySpan<char> partialCodePointBuffer);
+            combinedBuffer = combinedBuffer.Slice(0, ConcatInto(partialCodePointBuffer, value, combinedBuffer));
+
+            switch (Rune.DecodeFromUtf16(combinedBuffer, out _, out int charsConsumed))
```

Comment:
Question: what prevents us from always storing the partial code point/surrogate as UTF-8? Does the UTF-16 decoder fail if the char buffer starts with a low surrogate pair?

Reply:
Short answer: `Utf8.FromUtf16` will not decode partial surrogates (high or low). It will return NeedMoreData or InvalidData (depending on `isFinalBlock`) and will either leave the output empty or emit a replacement character. This is the same behavior in the other encoders/decoders.

Long answer: The UTF-16 low surrogate maps to the bottom 10 bits of the unicode scalar value and the high surrogate to the high 10 bits. In UTF-8 though, in a 4-byte code point (all high/low surrogate UTF-16 values need 4-bytes to represent in UTF-8), the 4th byte and 2 bits from the 3rd byte map to the bottom 10 bits of the unicode scalar value. And the remaining bits in byte 3 as well as bytes 1 and 2 map to the top 10 bits.

So byte 3 contains information for both the top 10 and bottom 10 bits of the Unicode scalar and transitively it contains information for both the high and low surrogates. So if we store the UTF-8 for a UTF-16, we would need to store 2 bytes and a few more bits. All of this is doable, but more complicated than just storing UTF-16 and adding a flag to indicate the value is UTF-16. And I'd guess decoders don't support this probably because they don't want to write the partial bytes, let alone the partial bits - instead they return NeedMoreData/InvalidOperation and write either nothing or a replacement char.